# Language Translation
In this project, you’re going to take a peek into the realm of neural network machine translation.  You’ll be training a sequence to sequence model on a dataset of English and French sentences that can translate new sentences from English to French.
## Get the Data
Since translating the whole language of English to French will take lots of time to train, we have provided you with a small portion of the English corpus.

In [1]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import helper
import problem_unittests as tests

source_path = 'data/small_vocab_en'
target_path = 'data/small_vocab_fr'
source_text = helper.load_data(source_path)
target_text = helper.load_data(target_path)

## Explore the Data
Play around with view_sentence_range to view different parts of the data.

In [2]:
view_sentence_range = (0, 10)

"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np

print('Dataset Stats')
print('Roughly the number of unique words: {}'.format(len({word: None for word in source_text.split()})))

sentences = source_text.split('\n')
word_counts = [len(sentence.split()) for sentence in sentences]
print('Number of sentences: {}'.format(len(sentences)))
print('Average number of words in a sentence: {}'.format(np.average(word_counts)))

print()
print('English sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(source_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))
print()
print('French sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(target_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))

Dataset Stats
Roughly the number of unique words: 227
Number of sentences: 137861
Average number of words in a sentence: 13.225277634719028

English sentences 0 to 10:
new jersey is sometimes quiet during autumn , and it is snowy in april .
the united states is usually chilly during july , and it is usually freezing in november .
california is usually quiet during march , and it is usually hot in june .
the united states is sometimes mild during june , and it is cold in september .
your least liked fruit is the grape , but my least liked is the apple .
his favorite fruit is the orange , but my favorite is the grape .
paris is relaxing during december , but it is usually chilly in july .
new jersey is busy during spring , and it is never hot in march .
our least liked fruit is the lemon , but my least liked is the grape .
the united states is sometimes busy during january , and it is sometimes warm in november .

French sentences 0 to 10:
new jersey est parfois calme pendant l' automne 

## Implement Preprocessing Function
### Text to Word Ids
As you did with other RNNs, you must turn the text into a number so the computer can understand it. In the function `text_to_ids()`, you'll turn `source_text` and `target_text` from words to ids.  However, you need to add the `<EOS>` word id at the end of `target_text`.  This will help the neural network predict when the sentence should end.

You can get the `<EOS>` word id by doing:
```python
target_vocab_to_int['<EOS>']
```
You can get other word ids using `source_vocab_to_int` and `target_vocab_to_int`.

In [3]:
def text_to_ids(source_text, target_text, source_vocab_to_int, target_vocab_to_int):
    """
    Convert source and target text to proper word ids
    :param source_text: String that contains all the source text.
    :param target_text: String that contains all the target text.
    :param source_vocab_to_int: Dictionary to go from the source words to an id
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: A tuple of lists (source_id_text, target_id_text)
    """
    # TODO: Implement Function
    source_text_ids = [[source_vocab_to_int.get(word, source_vocab_to_int['<UNK>']) for word in sentence.split(' ')] for sentence in source_text.split('\n')]
    target_text_ids = [[target_vocab_to_int.get(word, source_vocab_to_int['<UNK>']) for word in sentence.split(' ')] + [target_vocab_to_int['<EOS>']] for sentence in target_text.split('\n')]
    return source_text_ids, target_text_ids

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_text_to_ids(text_to_ids)

Tests Passed


### Preprocess all the data and save it
Running the code cell below will preprocess all the data and save it to file.

In [4]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
helper.preprocess_and_save_data(source_path, target_path, text_to_ids)

# Check Point
This is your first checkpoint. If you ever decide to come back to this notebook or have to restart the notebook, you can start from here. The preprocessed data has been saved to disk.

In [2]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np
import helper
import problem_unittests as tests

(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()

### Check the Version of TensorFlow and Access to GPU
This will check to make sure you have the correct version of TensorFlow and access to a GPU

In [3]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
from distutils.version import LooseVersion
import warnings
import tensorflow as tf
from tensorflow.python.layers.core import Dense

# Check TensorFlow Version
assert LooseVersion(tf.__version__) >= LooseVersion('1.1'), 'Please use TensorFlow version 1.1 or newer'
print('TensorFlow Version: {}'.format(tf.__version__))

# Check for a GPU
if not tf.test.gpu_device_name():
    warnings.warn('No GPU found. Please use a GPU to train your neural network.')
else:
    print('Default GPU Device: {}'.format(tf.test.gpu_device_name()))

TensorFlow Version: 1.1.0
Default GPU Device: /gpu:0


## Build the Neural Network
You'll build the components necessary to build a Sequence-to-Sequence model by implementing the following functions below:
- `model_inputs`
- `process_decoder_input`
- `encoding_layer`
- `decoding_layer_train`
- `decoding_layer_infer`
- `decoding_layer`
- `seq2seq_model`

### Input
Implement the `model_inputs()` function to create TF Placeholders for the Neural Network. It should create the following placeholders:

- Input text placeholder named "input" using the TF Placeholder name parameter with rank 2.
- Targets placeholder with rank 2.
- Learning rate placeholder with rank 0.
- Keep probability placeholder named "keep_prob" using the TF Placeholder name parameter with rank 0.
- Target sequence length placeholder named "target_sequence_length" with rank 1
- Max target sequence length tensor named "max_target_len" getting its value from applying tf.reduce_max on the target_sequence_length placeholder. Rank 0.
- Source sequence length placeholder named "source_sequence_length" with rank 1

Return the placeholders in the following the tuple (input, targets, learning rate, keep probability, target sequence length, max target sequence length, source sequence length)

In [4]:
def model_inputs():
    """
    Create TF Placeholders for input, targets, learning rate, and lengths of source and target sequences.
    :return: Tuple (input, targets, learning rate, keep probability, target sequence length,
    max target sequence length, source sequence length)
    """
    # TODO: Implement Function
    inputs = tf.placeholder(tf.int32, [None, None], name = 'input')
    targets = tf.placeholder(tf.int32, [None, None])
    learning_rate = tf.placeholder(tf.float32)
    keep_prob = tf.placeholder(tf.float32, name = 'keep_prob')
    target_sequence_length = tf.placeholder(tf.int32, [None,], name = 'target_sequence_length')
    max_target_sequence_length = tf.reduce_max(target_sequence_length, name = 'max_target_len')
    source_sequence_length = tf.placeholder(tf.int32, [None,], name = 'source_sequence_length')
    return (inputs, targets, learning_rate, keep_prob, target_sequence_length, max_target_sequence_length,
           source_sequence_length)


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_model_inputs(model_inputs)

Tests Passed


### Process Decoder Input
Implement `process_decoder_input` by removing the last word id from each batch in `target_data` and concat the GO ID to the begining of each batch.

In [5]:
def process_decoder_input(target_data, target_vocab_to_int, batch_size):
    """
    Preprocess target data for encoding
    :param target_data: Target Placehoder
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param batch_size: Batch Size
    :return: Preprocessed target data
    """
    # TODO: Implement Function
    ending = tf.strided_slice(target_data, [0, 0], [batch_size, -1], [1,1])
    target = tf.concat([tf.fill([batch_size, 1], target_vocab_to_int['<GO>']), ending], 1)
    return target

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_process_encoding_input(process_decoder_input)

Tests Passed


### Encoding
Implement `encoding_layer()` to create a Encoder RNN layer:
 * Embed the encoder input using [`tf.contrib.layers.embed_sequence`](https://www.tensorflow.org/api_docs/python/tf/contrib/layers/embed_sequence)
 * Construct a [stacked](https://github.com/tensorflow/tensorflow/blob/6947f65a374ebf29e74bb71e36fd82760056d82c/tensorflow/docs_src/tutorials/recurrent.md#stacking-multiple-lstms) [`tf.contrib.rnn.LSTMCell`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/LSTMCell) wrapped in a [`tf.contrib.rnn.DropoutWrapper`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/DropoutWrapper)
 * Pass cell and embedded input to [`tf.nn.dynamic_rnn()`](https://www.tensorflow.org/api_docs/python/tf/nn/dynamic_rnn)

In [7]:
from imp import reload
reload(tests)

def encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob, 
                   source_sequence_length, source_vocab_size, 
                   encoding_embedding_size):
    """
    Create encoding layer
    :param rnn_inputs: Inputs for the RNN
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param keep_prob: Dropout keep probability
    :param source_sequence_length: a list of the lengths of each sequence in the batch
    :param source_vocab_size: vocabulary size of source data
    :param encoding_embedding_size: embedding size of source data
    :return: tuple (RNN output, RNN state)
    """
    # TODO: Implement Function
    embedding = tf.contrib.layers.embed_sequence(rnn_inputs, vocab_size=source_vocab_size, embed_dim=encoding_embedding_size)
    def rnn_cell(rnn_size, keep_prob):
        cell = tf.contrib.rnn.LSTMCell(rnn_size, initializer=tf.random_uniform_initializer(-1, 1))
        cell = tf.contrib.rnn.DropoutWrapper(cell, output_keep_prob=keep_prob)
        return cell
    
    stacked_cell = tf.contrib.rnn.MultiRNNCell([rnn_cell(rnn_size, keep_prob) for _ in range(num_layers)])
    rnn_output, rnn_state = tf.nn.dynamic_rnn(stacked_cell, embedding, sequence_length=source_sequence_length, dtype=tf.float32)
    return rnn_output, rnn_state

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_encoding_layer(encoding_layer)

Tests Passed


### Decoding - Training
Create a training decoding layer:
* Create a [`tf.contrib.seq2seq.TrainingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/TrainingHelper) 
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [8]:

def decoding_layer_train(encoder_state, dec_cell, dec_embed_input, 
                         target_sequence_length, max_summary_length, 
                         output_layer, keep_prob):
    """
    Create a decoding layer for training
    :param encoder_state: Encoder State
    :param dec_cell: Decoder RNN Cell
    :param dec_embed_input: Decoder embedded input
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_summary_length: The length of the longest sequence in the batch
    :param output_layer: Function to apply the output layer
    :param keep_prob: Dropout keep probability
    :return: BasicDecoderOutput containing training logits and sample_id
    """
    # TODO: Implement Function
    dec_helper = tf.contrib.seq2seq.TrainingHelper(inputs=dec_embed_input, 
                                                   sequence_length=target_sequence_length,
                                                   time_major=False)
    dec_decoder = tf.contrib.seq2seq.BasicDecoder(dec_cell, dec_helper, encoder_state, output_layer=output_layer)
    final_outputs = tf.contrib.seq2seq.dynamic_decode(dec_decoder, impute_finished=True, 
                                                                   maximum_iterations=max_summary_length)[0]
    return final_outputs



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_train(decoding_layer_train)

Tests Passed


### Decoding - Inference
Create inference decoder:
* Create a [`tf.contrib.seq2seq.GreedyEmbeddingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/GreedyEmbeddingHelper)
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [9]:
def decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id,
                         end_of_sequence_id, max_target_sequence_length,
                         vocab_size, output_layer, batch_size, keep_prob):
    """
    Create a decoding layer for inference
    :param encoder_state: Encoder state
    :param dec_cell: Decoder RNN Cell
    :param dec_embeddings: Decoder embeddings
    :param start_of_sequence_id: GO ID
    :param end_of_sequence_id: EOS Id
    :param max_target_sequence_length: Maximum length of target sequences
    :param vocab_size: Size of decoder/target vocabulary
    :param decoding_scope: TensorFlow Variable Scope for decoding
    :param output_layer: Function to apply the output layer
    :param batch_size: Batch size
    :param keep_prob: Dropout keep probability
    :return: BasicDecoderOutput containing inference logits and sample_id
    """
    # TODO: Implement Function
    start_tokens = tf.tile(tf.constant([start_of_sequence_id], dtype=tf.int32), [batch_size], name='start_tokens')
    inf_helper = tf.contrib.seq2seq.GreedyEmbeddingHelper(dec_embeddings, start_tokens,
                                                         end_of_sequence_id)
    inf_decoder = tf.contrib.seq2seq.BasicDecoder(dec_cell, inf_helper, encoder_state, output_layer=output_layer)
    final_outputs = tf.contrib.seq2seq.dynamic_decode(inf_decoder, impute_finished=True,
                                                                  maximum_iterations=max_target_sequence_length)[0]
    return final_outputs



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_infer(decoding_layer_infer)

Tests Passed


### Build the Decoding Layer
Implement `decoding_layer()` to create a Decoder RNN layer.

* Embed the target sequences
* Construct the decoder LSTM cell (just like you constructed the encoder cell above)
* Create an output layer to map the outputs of the decoder to the elements of our vocabulary
* Use the your `decoding_layer_train(encoder_state, dec_cell, dec_embed_input, target_sequence_length, max_target_sequence_length, output_layer, keep_prob)` function to get the training logits.
* Use your `decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id, end_of_sequence_id, max_target_sequence_length, vocab_size, output_layer, batch_size, keep_prob)` function to get the inference logits.

Note: You'll need to use [tf.variable_scope](https://www.tensorflow.org/api_docs/python/tf/variable_scope) to share variables between training and inference.

In [10]:
def decoding_layer(dec_input, encoder_state,
                   target_sequence_length, max_target_sequence_length,
                   rnn_size,
                   num_layers, target_vocab_to_int, target_vocab_size,
                   batch_size, keep_prob, decoding_embedding_size):
    """
    Create decoding layer
    :param dec_input: Decoder input
    :param encoder_state: Encoder state
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_target_sequence_length: Maximum length of target sequences
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param target_vocab_size: Size of target vocabulary
    :param batch_size: The size of the batch
    :param keep_prob: Dropout keep probability
    :param decoding_embedding_size: Decoding embedding size
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """
    # TODO: Implement Function
    embedding_layer = tf.Variable(tf.random_uniform([target_vocab_size, decoding_embedding_size]))
    dec_inputs = tf.nn.embedding_lookup(embedding_layer, dec_input)
    
    def create_cell(rnn_size, keep_drop):
        cell = tf.contrib.rnn.LSTMCell(rnn_size, initializer=tf.random_uniform_initializer(-1, 1))
        #drop = tf.contrib.rnn.DropoutWrapper(cell, output_keep_prob=keep_prob)
        return cell
    
    stacked_cell = tf.contrib.rnn.MultiRNNCell([create_cell(rnn_size, keep_prob) for _ in range(num_layers)])
    
    output_layer = Dense(target_vocab_size, kernel_initializer=tf.truncated_normal_initializer(mean=0.0, stddev=0.1))
    
    with tf.variable_scope('decode'):
        training_logits = decoding_layer_train(encoder_state, stacked_cell, dec_inputs, target_sequence_length,
                             max_target_sequence_length, output_layer, keep_prob)
        
    with tf.variable_scope('decode', reuse=True):
        inference_logits = decoding_layer_infer(encoder_state, stacked_cell, embedding_layer, 
                                                target_vocab_to_int['<GO>'], target_vocab_to_int['<EOS>'],
                                                max_target_sequence_length, target_vocab_size, output_layer, 
                                                batch_size, keep_prob)
    return (training_logits, inference_logits)



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer(decoding_layer)

Tests Passed


### Build the Neural Network
Apply the functions you implemented above to:

- Encode the input using your `encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob,  source_sequence_length, source_vocab_size, encoding_embedding_size)`.
- Process target data using your `process_decoder_input(target_data, target_vocab_to_int, batch_size)` function.
- Decode the encoded input using your `decoding_layer(dec_input, enc_state, target_sequence_length, max_target_sentence_length, rnn_size, num_layers, target_vocab_to_int, target_vocab_size, batch_size, keep_prob, dec_embedding_size)` function.

In [11]:
def seq2seq_model(input_data, target_data, keep_prob, batch_size,
                  source_sequence_length, target_sequence_length,
                  max_target_sentence_length,
                  source_vocab_size, target_vocab_size,
                  enc_embedding_size, dec_embedding_size,
                  rnn_size, num_layers, target_vocab_to_int):
    """
    Build the Sequence-to-Sequence part of the neural network
    :param input_data: Input placeholder
    :param target_data: Target placeholder
    :param keep_prob: Dropout keep probability placeholder
    :param batch_size: Batch Size
    :param source_sequence_length: Sequence Lengths of source sequences in the batch
    :param target_sequence_length: Sequence Lengths of target sequences in the batch
    :param source_vocab_size: Source vocabulary size
    :param target_vocab_size: Target vocabulary size
    :param enc_embedding_size: Decoder embedding size
    :param dec_embedding_size: Encoder embedding size
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """
    # TODO: Implement Function
    _, enc_state =  encoding_layer(input_data, rnn_size, num_layers, keep_prob,  source_sequence_length,
                                source_vocab_size, enc_embedding_size)
    dec_input = process_decoder_input(target_data, target_vocab_to_int, batch_size)

    training_logits, inference_logits = decoding_layer(dec_input, enc_state, target_sequence_length, 
                                                       max_target_sentence_length, rnn_size, num_layers, 
                                                       target_vocab_to_int, target_vocab_size, batch_size, 
                                                       keep_prob, dec_embedding_size)
    return (training_logits, inference_logits)


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_seq2seq_model(seq2seq_model)

Tests Passed


## Neural Network Training
### Hyperparameters
Tune the following parameters:

- Set `epochs` to the number of epochs.
- Set `batch_size` to the batch size.
- Set `rnn_size` to the size of the RNNs.
- Set `num_layers` to the number of layers.
- Set `encoding_embedding_size` to the size of the embedding for the encoder.
- Set `decoding_embedding_size` to the size of the embedding for the decoder.
- Set `learning_rate` to the learning rate.
- Set `keep_probability` to the Dropout keep probability
- Set `display_step` to state how many steps between each debug output statement

In [5]:
# Number of Epochs
epochs = 30
# Batch Size
batch_size = 128
# RNN Size
rnn_size = 256
# Number of Layers
num_layers = 2
# Embedding Size
encoding_embedding_size = 64
decoding_embedding_size = 64
# Learning Rate
learning_rate = 0.01
# Dropout Keep Probability
keep_probability = 0.7
display_step = 50

### Build the Graph
Build the graph using the neural network you implemented.

In [13]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
save_path = 'checkpoints/dev'
(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()
max_target_sentence_length = max([len(sentence) for sentence in source_int_text])

train_graph = tf.Graph()
with train_graph.as_default():
    input_data, targets, lr, keep_prob, target_sequence_length, max_target_sequence_length, source_sequence_length = model_inputs()

    #sequence_length = tf.placeholder_with_default(max_target_sentence_length, None, name='sequence_length')
    input_shape = tf.shape(input_data)

    train_logits, inference_logits = seq2seq_model(tf.reverse(input_data, [-1]),
                                                   targets,
                                                   keep_prob,
                                                   batch_size,
                                                   source_sequence_length,
                                                   target_sequence_length,
                                                   max_target_sequence_length,
                                                   len(source_vocab_to_int),
                                                   len(target_vocab_to_int),
                                                   encoding_embedding_size,
                                                   decoding_embedding_size,
                                                   rnn_size,
                                                   num_layers,
                                                   target_vocab_to_int)


    training_logits = tf.identity(train_logits.rnn_output, name='logits')
    inference_logits = tf.identity(inference_logits.sample_id, name='predictions')

    masks = tf.sequence_mask(target_sequence_length, max_target_sequence_length, dtype=tf.float32, name='masks')

    with tf.name_scope("optimization"):
        # Loss function
        cost = tf.contrib.seq2seq.sequence_loss(
            training_logits,
            targets,
            masks)

        # Optimizer
        optimizer = tf.train.AdamOptimizer(lr)

        # Gradient Clipping
        gradients = optimizer.compute_gradients(cost)
        capped_gradients = [(tf.clip_by_value(grad, -1., 1.), var) for grad, var in gradients if grad is not None]
        train_op = optimizer.apply_gradients(capped_gradients)


Batch and pad the source and target sequences

In [14]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def pad_sentence_batch(sentence_batch, pad_int):
    """Pad sentences with <PAD> so that each sentence of a batch has the same length"""
    max_sentence = max([len(sentence) for sentence in sentence_batch])
    return [sentence + [pad_int] * (max_sentence - len(sentence)) for sentence in sentence_batch]


def get_batches(sources, targets, batch_size, source_pad_int, target_pad_int):
    """Batch targets, sources, and the lengths of their sentences together"""
    for batch_i in range(0, len(sources)//batch_size):
        start_i = batch_i * batch_size

        # Slice the right amount for the batch
        sources_batch = sources[start_i:start_i + batch_size]
        targets_batch = targets[start_i:start_i + batch_size]

        # Pad
        pad_sources_batch = np.array(pad_sentence_batch(sources_batch, source_pad_int))
        pad_targets_batch = np.array(pad_sentence_batch(targets_batch, target_pad_int))

        # Need the lengths for the _lengths parameters
        pad_targets_lengths = []
        for target in pad_targets_batch:
            pad_targets_lengths.append(len(target))

        pad_source_lengths = []
        for source in pad_sources_batch:
            pad_source_lengths.append(len(source))

        yield pad_sources_batch, pad_targets_batch, pad_source_lengths, pad_targets_lengths


### Train
Train the neural network on the preprocessed data. If you have a hard time getting a good loss, check the forms to see if anyone is having the same problem.

In [15]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def get_accuracy(target, logits):
    """
    Calculate accuracy
    """
    max_seq = max(target.shape[1], logits.shape[1])
    if max_seq - target.shape[1]:
        target = np.pad(
            target,
            [(0,0),(0,max_seq - target.shape[1])],
            'constant')
    if max_seq - logits.shape[1]:
        logits = np.pad(
            logits,
            [(0,0),(0,max_seq - logits.shape[1])],
            'constant')

    return np.mean(np.equal(target, logits))

# Split data to training and validation sets
train_source = source_int_text[batch_size:]
train_target = target_int_text[batch_size:]
valid_source = source_int_text[:batch_size]
valid_target = target_int_text[:batch_size]
(valid_sources_batch, valid_targets_batch, valid_sources_lengths, valid_targets_lengths ) = next(get_batches(valid_source,
                                                                                                             valid_target,
                                                                                                             batch_size,
                                                                                                             source_vocab_to_int['<PAD>'],
                                                                                                             target_vocab_to_int['<PAD>']))                                                                                                  
with tf.Session(graph=train_graph) as sess:
    sess.run(tf.global_variables_initializer())

    for epoch_i in range(epochs):
        for batch_i, (source_batch, target_batch, sources_lengths, targets_lengths) in enumerate(
                get_batches(train_source, train_target, batch_size,
                            source_vocab_to_int['<PAD>'],
                            target_vocab_to_int['<PAD>'])):
        
            _, loss = sess.run(
                [train_op, cost],
                {input_data: source_batch,
                 targets: target_batch,
                 lr: learning_rate,
                 target_sequence_length: targets_lengths,
                 source_sequence_length: sources_lengths,
                 keep_prob: keep_probability})


            if batch_i % display_step == 0 and batch_i > 0:


                batch_train_logits = sess.run(
                    inference_logits,
                    {input_data: source_batch,
                     source_sequence_length: sources_lengths,
                     target_sequence_length: targets_lengths,
                     keep_prob: 1.0})


                batch_valid_logits = sess.run(
                    inference_logits,
                    {input_data: valid_sources_batch,
                     source_sequence_length: valid_sources_lengths,
                     target_sequence_length: valid_targets_lengths,
                     keep_prob: 1.0})

                train_acc = get_accuracy(target_batch, batch_train_logits)

                valid_acc = get_accuracy(valid_targets_batch, batch_valid_logits)

                print('Epoch {:>3} Batch {:>4}/{} - Train Accuracy: {:>6.4f}, Validation Accuracy: {:>6.4f}, Loss: {:>6.4f}'
                      .format(epoch_i, batch_i, len(source_int_text) // batch_size, train_acc, valid_acc, loss))

    # Save Model
    saver = tf.train.Saver()
    saver.save(sess, save_path)
    print('Model Trained and Saved')

Epoch   0 Batch   50/1077 - Train Accuracy: 0.2414, Validation Accuracy: 0.3235, Loss: 3.3837
Epoch   0 Batch  100/1077 - Train Accuracy: 0.2629, Validation Accuracy: 0.3448, Loss: 2.8037
Epoch   0 Batch  150/1077 - Train Accuracy: 0.2872, Validation Accuracy: 0.3271, Loss: 2.1110
Epoch   0 Batch  200/1077 - Train Accuracy: 0.2895, Validation Accuracy: 0.3576, Loss: 1.6199
Epoch   0 Batch  250/1077 - Train Accuracy: 0.3803, Validation Accuracy: 0.4077, Loss: 1.1795
Epoch   0 Batch  300/1077 - Train Accuracy: 0.3277, Validation Accuracy: 0.4062, Loss: 1.2041
Epoch   0 Batch  350/1077 - Train Accuracy: 0.3488, Validation Accuracy: 0.4276, Loss: 1.0752
Epoch   0 Batch  400/1077 - Train Accuracy: 0.3734, Validation Accuracy: 0.4379, Loss: 0.9276
Epoch   0 Batch  450/1077 - Train Accuracy: 0.3117, Validation Accuracy: 0.4141, Loss: 0.9284
Epoch   0 Batch  500/1077 - Train Accuracy: 0.4156, Validation Accuracy: 0.4826, Loss: 0.8609
Epoch   0 Batch  550/1077 - Train Accuracy: 0.3766, Validati

Epoch   4 Batch  250/1077 - Train Accuracy: 0.7074, Validation Accuracy: 0.7454, Loss: 0.3100
Epoch   4 Batch  300/1077 - Train Accuracy: 0.7188, Validation Accuracy: 0.7475, Loss: 0.3240
Epoch   4 Batch  350/1077 - Train Accuracy: 0.7281, Validation Accuracy: 0.6719, Loss: 0.3176
Epoch   4 Batch  400/1077 - Train Accuracy: 0.7477, Validation Accuracy: 0.7116, Loss: 0.3289
Epoch   4 Batch  450/1077 - Train Accuracy: 0.7570, Validation Accuracy: 0.7259, Loss: 0.2929
Epoch   4 Batch  500/1077 - Train Accuracy: 0.7324, Validation Accuracy: 0.7504, Loss: 0.3009
Epoch   4 Batch  550/1077 - Train Accuracy: 0.7039, Validation Accuracy: 0.7280, Loss: 0.3631
Epoch   4 Batch  600/1077 - Train Accuracy: 0.7757, Validation Accuracy: 0.7617, Loss: 0.2911
Epoch   4 Batch  650/1077 - Train Accuracy: 0.7266, Validation Accuracy: 0.7322, Loss: 0.3052
Epoch   4 Batch  700/1077 - Train Accuracy: 0.7625, Validation Accuracy: 0.7362, Loss: 0.2965
Epoch   4 Batch  750/1077 - Train Accuracy: 0.7855, Validati

Epoch   8 Batch  450/1077 - Train Accuracy: 0.8750, Validation Accuracy: 0.8647, Loss: 0.1337
Epoch   8 Batch  500/1077 - Train Accuracy: 0.8934, Validation Accuracy: 0.8580, Loss: 0.1400
Epoch   8 Batch  550/1077 - Train Accuracy: 0.8477, Validation Accuracy: 0.8466, Loss: 0.1455
Epoch   8 Batch  600/1077 - Train Accuracy: 0.8858, Validation Accuracy: 0.8679, Loss: 0.1309
Epoch   8 Batch  650/1077 - Train Accuracy: 0.8633, Validation Accuracy: 0.8821, Loss: 0.1389
Epoch   8 Batch  700/1077 - Train Accuracy: 0.8902, Validation Accuracy: 0.8675, Loss: 0.1369
Epoch   8 Batch  750/1077 - Train Accuracy: 0.8750, Validation Accuracy: 0.8509, Loss: 0.1459
Epoch   8 Batch  800/1077 - Train Accuracy: 0.8836, Validation Accuracy: 0.8722, Loss: 0.1314
Epoch   8 Batch  850/1077 - Train Accuracy: 0.8583, Validation Accuracy: 0.8498, Loss: 0.1479
Epoch   8 Batch  900/1077 - Train Accuracy: 0.8848, Validation Accuracy: 0.8594, Loss: 0.1587
Epoch   8 Batch  950/1077 - Train Accuracy: 0.8821, Validati

Epoch  12 Batch  650/1077 - Train Accuracy: 0.9059, Validation Accuracy: 0.8881, Loss: 0.1259
Epoch  12 Batch  700/1077 - Train Accuracy: 0.8945, Validation Accuracy: 0.9034, Loss: 0.1242
Epoch  12 Batch  750/1077 - Train Accuracy: 0.8113, Validation Accuracy: 0.7390, Loss: 0.2335
Epoch  12 Batch  800/1077 - Train Accuracy: 0.8793, Validation Accuracy: 0.8459, Loss: 0.1488
Epoch  12 Batch  850/1077 - Train Accuracy: 0.8709, Validation Accuracy: 0.8732, Loss: 0.1720
Epoch  12 Batch  900/1077 - Train Accuracy: 0.8887, Validation Accuracy: 0.8707, Loss: 0.1495
Epoch  12 Batch  950/1077 - Train Accuracy: 0.8899, Validation Accuracy: 0.8732, Loss: 0.1110
Epoch  12 Batch 1000/1077 - Train Accuracy: 0.9022, Validation Accuracy: 0.9059, Loss: 0.1027
Epoch  12 Batch 1050/1077 - Train Accuracy: 0.8922, Validation Accuracy: 0.9034, Loss: 0.1242
Epoch  13 Batch   50/1077 - Train Accuracy: 0.9086, Validation Accuracy: 0.8800, Loss: 0.1147
Epoch  13 Batch  100/1077 - Train Accuracy: 0.9129, Validati

Epoch  16 Batch  850/1077 - Train Accuracy: 0.8962, Validation Accuracy: 0.9237, Loss: 0.1212
Epoch  16 Batch  900/1077 - Train Accuracy: 0.9254, Validation Accuracy: 0.9208, Loss: 0.1047
Epoch  16 Batch  950/1077 - Train Accuracy: 0.9252, Validation Accuracy: 0.9023, Loss: 0.0864
Epoch  16 Batch 1000/1077 - Train Accuracy: 0.9364, Validation Accuracy: 0.9059, Loss: 0.0811
Epoch  16 Batch 1050/1077 - Train Accuracy: 0.9336, Validation Accuracy: 0.8846, Loss: 0.1245
Epoch  17 Batch   50/1077 - Train Accuracy: 0.9281, Validation Accuracy: 0.8842, Loss: 0.1101
Epoch  17 Batch  100/1077 - Train Accuracy: 0.9066, Validation Accuracy: 0.9041, Loss: 0.1060
Epoch  17 Batch  150/1077 - Train Accuracy: 0.9275, Validation Accuracy: 0.8899, Loss: 0.0971
Epoch  17 Batch  200/1077 - Train Accuracy: 0.9156, Validation Accuracy: 0.9016, Loss: 0.1074
Epoch  17 Batch  250/1077 - Train Accuracy: 0.9286, Validation Accuracy: 0.9023, Loss: 0.0924
Epoch  17 Batch  300/1077 - Train Accuracy: 0.9359, Validati

Epoch  20 Batch 1050/1077 - Train Accuracy: 0.9539, Validation Accuracy: 0.9371, Loss: 0.0746
Epoch  21 Batch   50/1077 - Train Accuracy: 0.9418, Validation Accuracy: 0.9141, Loss: 0.0817
Epoch  21 Batch  100/1077 - Train Accuracy: 0.9203, Validation Accuracy: 0.9201, Loss: 0.0861
Epoch  21 Batch  150/1077 - Train Accuracy: 0.9449, Validation Accuracy: 0.9318, Loss: 0.0778
Epoch  21 Batch  200/1077 - Train Accuracy: 0.9344, Validation Accuracy: 0.9379, Loss: 0.0871
Epoch  21 Batch  250/1077 - Train Accuracy: 0.9286, Validation Accuracy: 0.9350, Loss: 0.0837
Epoch  21 Batch  300/1077 - Train Accuracy: 0.9618, Validation Accuracy: 0.9336, Loss: 0.0741
Epoch  21 Batch  350/1077 - Train Accuracy: 0.9309, Validation Accuracy: 0.9205, Loss: 0.0707
Epoch  21 Batch  400/1077 - Train Accuracy: 0.9332, Validation Accuracy: 0.9435, Loss: 0.0801
Epoch  21 Batch  450/1077 - Train Accuracy: 0.9289, Validation Accuracy: 0.9201, Loss: 0.0763
Epoch  21 Batch  500/1077 - Train Accuracy: 0.9254, Validati

Epoch  25 Batch  200/1077 - Train Accuracy: 0.9430, Validation Accuracy: 0.9290, Loss: 0.0836
Epoch  25 Batch  250/1077 - Train Accuracy: 0.9293, Validation Accuracy: 0.9371, Loss: 0.0824
Epoch  25 Batch  300/1077 - Train Accuracy: 0.9572, Validation Accuracy: 0.9538, Loss: 0.0630
Epoch  25 Batch  350/1077 - Train Accuracy: 0.9344, Validation Accuracy: 0.9407, Loss: 0.0638
Epoch  25 Batch  400/1077 - Train Accuracy: 0.9242, Validation Accuracy: 0.9308, Loss: 0.0852
Epoch  25 Batch  450/1077 - Train Accuracy: 0.9281, Validation Accuracy: 0.8956, Loss: 0.0734
Epoch  25 Batch  500/1077 - Train Accuracy: 0.9469, Validation Accuracy: 0.9442, Loss: 0.0768
Epoch  25 Batch  550/1077 - Train Accuracy: 0.9422, Validation Accuracy: 0.9407, Loss: 0.0692
Epoch  25 Batch  600/1077 - Train Accuracy: 0.9457, Validation Accuracy: 0.9304, Loss: 0.0636
Epoch  25 Batch  650/1077 - Train Accuracy: 0.9293, Validation Accuracy: 0.9162, Loss: 0.0768
Epoch  25 Batch  700/1077 - Train Accuracy: 0.9414, Validati

Epoch  29 Batch  400/1077 - Train Accuracy: 0.9230, Validation Accuracy: 0.9482, Loss: 0.0837
Epoch  29 Batch  450/1077 - Train Accuracy: 0.9199, Validation Accuracy: 0.9197, Loss: 0.0696
Epoch  29 Batch  500/1077 - Train Accuracy: 0.9441, Validation Accuracy: 0.9205, Loss: 0.0664
Epoch  29 Batch  550/1077 - Train Accuracy: 0.9301, Validation Accuracy: 0.9176, Loss: 0.0683
Epoch  29 Batch  600/1077 - Train Accuracy: 0.9312, Validation Accuracy: 0.9247, Loss: 0.0606
Epoch  29 Batch  650/1077 - Train Accuracy: 0.9461, Validation Accuracy: 0.9176, Loss: 0.0679
Epoch  29 Batch  700/1077 - Train Accuracy: 0.9496, Validation Accuracy: 0.9506, Loss: 0.0604
Epoch  29 Batch  750/1077 - Train Accuracy: 0.9371, Validation Accuracy: 0.9414, Loss: 0.0762
Epoch  29 Batch  800/1077 - Train Accuracy: 0.9316, Validation Accuracy: 0.9190, Loss: 0.0657
Epoch  29 Batch  850/1077 - Train Accuracy: 0.9278, Validation Accuracy: 0.9496, Loss: 0.0873
Epoch  29 Batch  900/1077 - Train Accuracy: 0.9121, Validati

### Save Parameters
Save the `batch_size` and `save_path` parameters for inference.

In [16]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# Save parameters for checkpoint
helper.save_params(save_path)

# Checkpoint

In [1]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import tensorflow as tf
import numpy as np
import helper
import problem_unittests as tests

_, (source_vocab_to_int, target_vocab_to_int), (source_int_to_vocab, target_int_to_vocab) = helper.load_preprocess()
load_path = helper.load_params()

## Sentence to Sequence
To feed a sentence into the model for translation, you first need to preprocess it.  Implement the function `sentence_to_seq()` to preprocess new sentences.

- Convert the sentence to lowercase
- Convert words into ids using `vocab_to_int`
 - Convert words not in the vocabulary, to the `<UNK>` word id.

In [3]:
def sentence_to_seq(sentence, vocab_to_int):
    """
    Convert a sentence to a sequence of ids
    :param sentence: String
    :param vocab_to_int: Dictionary to go from the words to an id
    :return: List of word ids
    """
    # TODO: Implement Function
    sentence = sentence.lower()
    sentence_ids = [vocab_to_int.get(word, vocab_to_int['<UNK>']) for word in sentence.split(' ')]
    return sentence_ids


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_sentence_to_seq(sentence_to_seq)

Tests Passed


## Translate
This will translate `translate_sentence` from English to French.

In [11]:
translate_sentence = 'he saw a old yellow truck .'


"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
translate_sentence = sentence_to_seq(translate_sentence, source_vocab_to_int)

loaded_graph = tf.Graph()
with tf.Session(graph=loaded_graph) as sess:
    # Load saved model
    loader = tf.train.import_meta_graph(load_path + '.meta')
    loader.restore(sess, load_path)

    input_data = loaded_graph.get_tensor_by_name('input:0')
    logits = loaded_graph.get_tensor_by_name('predictions:0')
    target_sequence_length = loaded_graph.get_tensor_by_name('target_sequence_length:0')
    source_sequence_length = loaded_graph.get_tensor_by_name('source_sequence_length:0')
    keep_prob = loaded_graph.get_tensor_by_name('keep_prob:0')

    translate_logits = sess.run(logits, {input_data: [translate_sentence]*batch_size,
                                         target_sequence_length: [len(translate_sentence)*2]*batch_size,
                                         source_sequence_length: [len(translate_sentence)]*batch_size,
                                         keep_prob: 1.0})[0]

print('Input')
print('  Word Ids:      {}'.format([i for i in translate_sentence]))
print('  English Words: {}'.format([source_int_to_vocab[i] for i in translate_sentence]))

print('\nPrediction')
print('  Word Ids:      {}'.format([i for i in translate_logits]))
print('  French Words: {}'.format(" ".join([target_int_to_vocab[i] for i in translate_logits])))


INFO:tensorflow:Restoring parameters from checkpoints/dev
Input
  Word Ids:      [226, 127, 59, 20, 121, 183, 6]
  English Words: ['he', 'saw', 'a', 'old', 'yellow', 'truck', '.']

Prediction
  Word Ids:      [40, 23, 134, 253, 325, 337, 6, 1]
  French Words: il envisage de traduire les jaune . <EOS>


## Imperfect Translation
You might notice that some sentences translate better than others.  Since the dataset you're using only has a vocabulary of 227 English words of the thousands that you use, you're only going to see good results using these words.  For this project, you don't need a perfect translation. However, if you want to create a better translation model, you'll need better data.

You can train on the [WMT10 French-English corpus](http://www.statmt.org/wmt10/training-giga-fren.tar).  This dataset has more vocabulary and richer in topics discussed.  However, this will take you days to train, so make sure you've a GPU and the neural network is performing well on dataset we provided.  Just make sure you play with the WMT10 corpus after you've submitted this project.
## Submitting This Project
When submitting this project, make sure to run all the cells before saving the notebook. Save the notebook file as "dlnd_language_translation.ipynb" and save it as a HTML file under "File" -> "Download as". Include the "helper.py" and "problem_unittests.py" files in your submission.